### PR TITLE
[SofaDeformable] Bind LinearSpring

### PR DESCRIPTION
**Pull request by ScheiklP**
#77

### CODE REVIEW

**Comment by jnbrunet** [comment]
Hey @ScheiklP ,

Thanks again for this PR.

You got a compilation error here since the `LinearSpring` class does not inherit from `sofa::core::objectmodel::Base`, hence its memory allocation/release is not managed by `boost::intrusive_ptr` (unlike all SOFA components that derive from Base). 

In SofaPython3, to indicate to pybind11 that we are managing the pointers ourselves (a concept called [type holders](https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#custom-smart-pointers) in pybind11), we use `sofapython3::py_shared_ptr` which is just an alias for `boost::intrusive_ptr`.

Hence, to bind a SOFA class that inherits from Base, we do:

```c++
py::class_<MyComponent, sofa::core::objectmodel::Base, py_shared_ptr<MyComponent>> s (m, "MyComponent");
``` 

In your case, the LinearSpring does not inherit from Base, so simply remove the `py_shared_ptr` holder type. 

**TLDR**:
Change
```c++
py::class_<LinearSpringR, py_shared_ptr<LinearSpringR>> s (m, "LinearSpring", sofapython3::doc::SofaDeformable::LinearSpringClass);
```
to
```c++
py::class_<LinearSpringR> s (m, "LinearSpring", sofapython3::doc::SofaDeformable::LinearSpringClass);
```

J-N

File: bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    - jean-nicolas.brunet@inria.fr
+    - thierry.gaugry@inria.fr
+    - paul.scheikl@kit.edu
+********************************************************************/
+
+#include <pybind11/pybind11.h>
+#include "Binding_LinearSpring.h"
+#include "Binding_LinearSpring_doc.h"
+
+#include <SofaDeformable/config.h>
```
Comment (jnbrunet) [inline bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp:33]: > #include <SofaDeformable/config.h>

Not needed

File: bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#include <SofaDeformable/SpringForceField.h>
+
+/* #include <SofaPython3/Sofa/Core/Binding_Base.h> */
```
Comment (jnbrunet) [inline bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp:36]: > /* #include <SofaPython3/Sofa/Core/Binding_Base.h> */

Remove commented code

File: bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+#include "../../../../Sofa/src/SofaPython3/Sofa/Core/Binding_Base.h"
```
Comment (jnbrunet) [inline bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp:37]: > #include "../../../../Sofa/src/SofaPython3/Sofa/Core/Binding_Base.h"

`#include <SofaPython3/Sofa/Core/Binding_Base.h>`

File: bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#include <sofa/config.h>
+#include <SofaPython3/PythonFactory.h>
```
Comment (jnbrunet) [inline bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp:39]: >#include <sofa/config.h>
#include <SofaPython3/PythonFactory.h>

Not needed

File: bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+namespace sofapython3 {
+
+using sofa::component::interactionforcefield::LinearSpring;
+typedef LinearSpring<SReal> LinearSpringR;
+
+namespace py { using namespace pybind11; }
+
+void moduleAddLinearSpring(pybind11::module& m) {
+    py::class_<LinearSpringR, py_shared_ptr<LinearSpringR>> s (m, "LinearSpring", sofapython3::doc::SofaDeformable::LinearSpringClass);
+    s.def(py::init<sofa::Index,
```
Comment (jnbrunet) [inline bindings/Modules/src/SofaPython3/SofaDeformable/Binding_LinearSpring.cpp:51]: It could be useful here to add [class fields](https://pybind11.readthedocs.io/en/stable/classes.html#instance-and-static-fields)

**Comment by ScheiklP** [comment]
> Hey @ScheiklP ,
> 
> Thanks again for this PR.
> 
> You got a compilation error here since the `LinearSpring` class does not inherit from `sofa::core::objectmodel::Base`, hence its memory allocation/release is not managed by `boost::intrusive_ptr` (unlike all SOFA components that derive from Base).
> 
> In SofaPython3, to indicate to pybind11 that we are managing the pointers ourselves (a concept called [type holders](https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#custom-smart-pointers) in pybind11), we use `sofapython3::py_shared_ptr` which is just an alias for `boost::intrusive_ptr`.
> 
> Hence, to bind a SOFA class that inherits from Base, we do:
> 
> ```c++
> py::class_<MyComponent, sofa::core::objectmodel::Base, py_shared_ptr<MyComponent>> s (m, "MyComponent");
> ```
> 
> In your case, the LinearSpring does not inherit from Base, so simply remove the `py_shared_ptr` holder type.
> 
> **TLDR**:
> Change
> 
> ```c++
> py::class_<LinearSpringR, py_shared_ptr<LinearSpringR>> s (m, "LinearSpring", sofapython3::doc::SofaDeformable::LinearSpringClass);
> ```
> 
> to
> 
> ```c++
> py::class_<LinearSpringR> s (m, "LinearSpring", sofapython3::doc::SofaDeformable::LinearSpringClass);
> ```
> 
> J-N

Thanks a lot for that explanation! I did not realize that that was a smart pointer thing. :)

**Comment by ScheiklP** [comment]
LinearSpring binds fine!

The SpringForceField bindings build as well, but downcasting is not working yet. 
``` 
FF = node1.addObject(
        "SpringForceField",
        template="Vec3d",
        name="SpringForceField",
        object1=node1.collision.getLinkPath(),
        object2=node2.collision.getLinkPath(),
        spring="0 0 1 5 5"
    ) # works well -> spring appears in SofaGui

    node1.init()
    print(FF) # -> <Sofa.Core.Object object at 0x7f0264069f50>
    FF.clear() # -> [ERROR]   [SofaPython3::SceneLoader] KeyError: 'clear'
```

```
from Sofa.SofaDeformable import SpringForceField
````
Results in 
```
[ERROR]   [SofaPython3::SceneLoader] ImportError: /home/paul/repos/sofa/build/v2012/lib/python3/site-packages/Sofa/SofaDeformable.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZN11sofapython325moduleAddSpringForceFieldIN4sofa11defaulttype14StdVectorTypesINS2_3VecILj3EdEES5_dEEEEvRN8pybind116moduleE

# Demangled:
void sofapython3::moduleAddSpringForceField<sofa::defaulttype::StdVectorTypes<sofa::defaulttype::Vec<3u, double>, sofa::defaulttype::Vec<3u, double>, double> >(pybind11::module&)
```

**Comment by jnbrunet** [comment]
Hey @ScheiklP ,

You got this error since the templated function `moduleAddSpringForceField<T>` is defined inside a cpp file (_Binding_SpringForceField.cpp_), and you are instantiating the template values `moduleAddSpringForceField<Vec3dTypes>` and  `moduleAddSpringForceField<Vec6dTypes>` inside another cpp file (_Module_SofaDeformable.cpp_) which doesn't have access to the definition (it isn't inside a .h or .inl file).

I've made a very quick fix here, which should get you as far as correctly loading the `Sofa.SofaDeformable` python module:

https://github.com/jnbrunet/SofaPython3/commit/9e41dd51755943e97f34890dfa9554c64587474c

However, it looks like getting the value of `StiffSpringForceField.spring` still return an array of empties. I will try to have a better look next week. Until then, you can already play with it a little bit and maybe, for example, bind a helper function that will delete some springs based on a given set of indices.

J-N

**Comment by ScheiklP** [comment]
Hi @jnbrunet ,
ah damn, of course :man_facepalming: ... thanks a lot! :)

Okay cool, I will try that. :)

**Comment by ScheiklP** [comment]
Hi @jnbrunet ,
That worked surprisingly well!

`StiffSpringForceField.spring` is still a list of empties, but SpringForceField.getSprings() is a fine replacement for now.

And I noticed a weird behavior. Down casting only works, when the module is actually imported in python.
Without import:
```
FF = node1.addObject(
        "SpringForceField",
        template="Vec3d",
        name="SpringForceField",
        object1=node1.collision.getLinkPath(),
        object2=node2.collision.getLinkPath(),
        spring="0 0 1 5 5"
    )

    node1.init()
    print(FF) # -> <Sofa.Core.Object object at 0x7f0264069f50>
    FF.clear() # -> [ERROR]   [SofaPython3::SceneLoader] KeyError: 'clear'
```

With import:
```
import Sofa.SofaDeformable

FF = node1.addObject(
        "SpringForceField",
        template="Vec3d",
        name="SpringForceField",
        object1=node1.collision.getLinkPath(),
        object2=node2.collision.getLinkPath(),
        spring="0 0 1 5 5"
    )

    node1.init()
    print(FF) # -> <Sofa.SofaDeformable.SpringForceFieldStdVectorTypesVec<3u,double>,Vec<3u,double>,double>> object at 0x7f5d8e8a18f0>
    FF.clear() # works
```

Cheers,
Paul

**Comment by jnbrunet** [comment]
Awesome!

> And I noticed a weird behavior. Down casting only works, when the module is actually imported in python.

That is actually the correct behavior. The bindings are tied to the binary library `libSofaDeformable.cython.(...).so`. When you do `import SofaDeformable`, python will load this library and call this method defined in your _Binding_SpringForceField.cpp_  
```c++
PYBIND11_MODULE(SofaDeformable, m) {
(...)
}
```

Which will then call your bindings and register them. Now, only registering them through pybind11 (using `py::class_<>` and other) would work if some c++ method returns a `SpringForceField` pointer. In this case, pybind11 would correctly convert the object to a python instance, and vice versa. However, when you do
```python
o = root.addObject('SpringForceField', template='vec3d')
```
The `addObject` method goes through the SOFA's object manager, which will look at the object name `SpringForceField` and the template name 'vec3d' and create the appropriate instance (`SpringForceField<Vec3D>`). It will then return a `sofa::core::Base` pointer. So in this case, pybind11 has no way to tell that this `Base *` is actually a `SpringForceField<Vec3D> *`. This is where our SofaPython3's `PythonFactory` comes into play. It will use the type introspection of SOFA to determine the name of the class (and its template). Then, it will use the callback that you specified in your binding code:

```c++
    // register the binding in the downcasting subsystem
    PythonFactory::registerType<SpringForceField>([](sofa::core::objectmodel::Base* object){
        return py::cast(dynamic_cast<SpringForceField*>(object));
    });
``` 

Again, all of this is "set up" when you do `import SofaDeformable`.

Let me know if this is still unclear, or if you have other questions. This discussion could be useful for others (and we should really find some time to put this in the documentation ! ).

J-N

**Comment by ScheiklP** [comment]
That makes a lot of sense, thanks! :) 

Definitely! Would you want to do this in the official https://www.sofa-framework.org/ or rather the sphinx documentation in the repository?

**Comment by jnbrunet** [comment]
For now I think we will continue with the sphinx documentation, and maybe at some point we will move it to the SOFA official website.

**Comment by jnbrunet** [comment]
Hey @ScheiklP 

Thanks again for this PR !

I just finished reviewing it. I added two commit:
1. f06c7f62b8349f045e463ba6d1b73ffa5276b0b6 Merge the master branch
2. fd0fa1950e56fc37ef283cc85dd35bafc79542fc The following changes:
    1. Some small cleanups
    2. I added some tests for the bindings of SpringForceField
    3. I added the tests of LinearSpring and SpringForceField to the CI
    4. The `SpringForceField.removeSprings(indices)` was misleading. Since it was deleting the springs one by one, after each deletion, the indices would have to be decreased by one in order to work as one would expect. For example, if I had 5 springs, and expected that `SpringForceField.removeSprings([3,4])` deletes the last two springs, it would do:
        a. Indices of the springs are [0, 1, 2, 3, 4]
        b. Call `SpringForceField.removeSpring(3)` -> Indices are now [0, 1, 2, 3]
        c. Call `SpringForceField.removeSpring(4)` -> Index 4 doesn't exists

    Hence, to make it work, I changed the binding of `SpringForceField.removeSpring` so that it reverse-sort the indices before removing them.

Let me know if your are ok with these changes, in which case I will squash your PR into master.

J-N

**Comment by ScheiklP** [comment]
Hi @jnbrunet ,
and thank you very much for guiding me through the process! :) 

Oh wow, good catch, I did not notice that the indices get updated after removing a spring!

The changes look perfect. :)

Cheers,
Paul
